subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/PublishedDependencyConstraintsIntegrationTest.groovy

```diff
@@ -0,0 +1,401 @@
+                    expectGetArtifact()
+                }
+            }
+            'org:root:1.0' {
+                expectGetMetadata()
+                expectGetArtifact()
+            }
+        }
+
+        when:
+        run 'checkDeps'
+
+        then:
+        resolve.expectGraph {
+            root(":", ":test:") {
+                module("org:root:1.0") {
+                    if (available) {
+                        module("org:foo:1.1")
+                    }
+                }
+                if (available) {
+                    edge("org:foo:1.0","org:foo:1.1").byConflictResolution()
+                } else {
+                    module("org:foo:1.0")
+                }
+            }
+        }
+    }
+
+    void "dependency constraint is included into the result of resolution when a hard dependency is also added transitively"() {
+        given:
+        def available = featureAvailable()
+        repository {
+            'org:foo:1.0'()
+            'org:foo:1.1'()
+            'org:root1:1.0' {
+                constraint 'org:foo:1.1'
+            }
+            'org:root2:1.0' {
+                dependsOn 'org:foo:1.0'
+            }
+        }
+
+        buildFile << """
+            dependencies {
+                conf 'org:root1:1.0'
+                conf 'org:root2:1.0'
+            }
+        """
+
+        repositoryInteractions {
+            'org:foo:1.0' {
+                expectGetMetadata()
+                if (!available) {
+                    expectGetArtifact()
+                }
+            }
+            'org:foo:1.1' {
+                if (available) {
+                    expectGetMetadata()
+                    expectGetArtifact()
+                }
+            }
+            'org:root1:1.0' {
+                expectGetMetadata()
+                expectGetArtifact()
+            }
+            'org:root2:1.0' {
+                expectGetMetadata()
+                expectGetArtifact()
+            }
+        }
+
+        when:
+        run 'checkDeps'
+
+        then:
+        resolve.expectGraph {
+            root(":", ":test:") {
+                module("org:root1:1.0") {
+                    if (available) {
+                        module("org:foo:1.1")
+                    }
+                }
+                module("org:root2:1.0") {
+                    if (available) {
+                        edge("org:foo:1.0","org:foo:1.1").byConflictResolution()
+                    } else {
+                        module("org:foo:1.0")
+                    }
+                }
+            }
+        }
+    }
+
+    void "range resolution kicks in with optional dependencies"() {
+        given:
+        def available = featureAvailable()
+        repository {
+            'org:foo:1.0'()
+            'org:foo:1.1'()
+            'org:foo:1.2'()
+            'org:bar:1.0' {
+                dependsOn 'org:foo:[1.1,1.2]'
+            }
+            'org:root:1.0' {
```

Comment:
I think the name 'root' is confusing here. Same in the previous test

Reply:
Renamed it to `first-level`